core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java

```diff
@@ -0,0 +1,191 @@
+        List<Experiment> experiments = projectConfig.getExperiments();
+        if(experiments != null) {
+            for (Experiment experiment : experiments) {
+                OptimizelyExperiment optimizelyExperiment = new OptimizelyExperiment(
+                    experiment.getId(),
+                    experiment.getKey(),
+                    getVariationsMap(projectConfig, experiment.getVariations(), experiment.getId())
+                );
+                experimentsMap.put(experiment.getKey(), optimizelyExperiment);            
+            }
+        }
+        return experimentsMap;
+    }
+
+    private Map<String, OptimizelyVariation> getVariationsMap(ProjectConfig projectConfig, List<Variation> variations, String experimentId) {
+        Map<String, OptimizelyVariation> variationsMap = new HashMap<>();
+        if (variations != null) {
+            for (Variation variation : variations) {
+                List<String> features = projectConfig.getExperimentFeatureKeyMapping().get(experimentId);
+                Boolean isFeatureEnabled = features != null && !features.isEmpty() ? variation.getFeatureEnabled() : null;
+                OptimizelyVariation optimizelyVariation = new OptimizelyVariation(
+                    variation.getId(),
+                    variation.getKey(),
+                    isFeatureEnabled,
+                    getMergedVariablesMap(projectConfig, variation, experimentId)
+                );
+                variationsMap.put(variation.getKey(), optimizelyVariation);
+            }
+        }
+        return variationsMap;
+    }
+
+    /**
+     * Merges feature key and type from feature variables to variation variables
+     *
+     * @param projectConfig             the current projectConfig
+     * @param variation                 variation for which the value of {@link FeatureVariable} needs to be merged
+     * @param experimentId              {@link Experiment} id for getting its feature id
+     * @param featureVariablesMap       map for {@link FeatureVariable} of a {@link FeatureFlag}
+     * @return optimizelyVariableMap    returns the map after merging the value of {@link FeatureVariable}
+     */
+    private Map<String, OptimizelyVariable> getMergedVariablesMap(ProjectConfig projectConfig, Variation variation, String experimentId) {
+
+        Map<String, OptimizelyVariable> optimizelyVariableMap = new HashMap<>();
+        Map<String, List<FeatureVariable>> featureIDVariablesMap = new HashMap<>();
+        
+        List<String> featureKeys = projectConfig.getExperimentFeatureKeyMapping().get(experimentId);
+        String featureKey = featureKeys != null ? featureKeys.get(0) : null;
+
+        List<FeatureFlag> featureFlags = projectConfig.getFeatureFlags();
+        if (featureFlags != null) {
+            for (FeatureFlag featureFlag : featureFlags) {
+                featureIDVariablesMap.put(featureFlag.getKey(), featureFlag.getVariables());
+            }
+        }
```

Comment:
We do not need to get variables for all featureFlags - just getVariables() for the target featureKey.